Add linked-on-or-after quirk on execution timing change of module script.

basuke opened the pull request:
#### dcd4aaf9fdc6396ca6cf351d36acab6b65852dfd
<pre>
Add linked-on-or-after quirk on execution timing change of module script.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300905">https://bugs.webkit.org/show_bug.cgi?id=300905</a>
<a href="https://rdar.apple.com/162377646">rdar://162377646</a>

Reviewed by Tim Nguyen.

Execution timing of module scripts are changed to match the HTML spec on 301345@main.
To protect the clients of WKWebView, add linked-on-or-after quirk to prevent the behavior
change.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::canChangeModuleScriptsExecutionTiming):
(WebCore::HTMLDocumentParser::notifyFinished):

Canonical link: <a href="https://commits.webkit.org/301665@main">https://commits.webkit.org/301665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70fd729de9b4b7d9dd6e7951e8fe7cbb58c894c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46311 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37274 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133632 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78321 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/92f2150e-bb84-485d-a4d1-98d18787c073) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46943 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54849 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3b723d15-4ded-4c11-aa32-91324ec1f034) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129615 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/37562 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113298 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/76914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d4147855-9e30-4529-8610-783a100c8a33) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31479 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77028 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118716 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136205 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125132 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41049 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/104908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109653 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/104609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26677 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50101 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28436 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50788 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53276 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59084 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158176 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52548 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39576 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55884 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54288 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/parser/HTMLDocumentParser.cpp:
```diff
@@ -582,6 +586,17 @@ void HTMLDocumentParser::appendCurrentInputStreamToPreloadScannerAndScan()
+#if PLATFORM(COCOA)
+    // https://bugs.webkit.org/show_bug.cgi?id=300905
+    static bool caChange = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ExecutionTimingChangeOfModuleScripts);
+    return caChange;
```
```suggestion
    static const bool canChangeTiming = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ExecutionTimingChangeOfModuleScripts);
    return canChangeTiming;
```

nt1m commented on Source/WebCore/html/parser/HTMLDocumentParser.cpp:
```diff
@@ -582,6 +586,17 @@ void HTMLDocumentParser::appendCurrentInputStreamToPreloadScannerAndScan()
+#if PLATFORM(COCOA)
+    // https://bugs.webkit.org/show_bug.cgi?id=300905
+    static const bool caChange = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ExecutionTimingChangeOfModuleScripts);
+    return caChange;
```
```suggestion
    static const bool canChangeTiming = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ExecutionTimingChangeOfModuleScripts);
    return canChangeTiming;
```

basuke committed:
Add linked-on-or-after quirk on execution timing change of module script.
https://bugs.webkit.org/show_bug.cgi?id=300905
rdar://162377646

Reviewed by Tim Nguyen.

Execution timing of module scripts are changed to match the HTML spec on 301345@main.
To protect the clients of WKWebView, add linked-on-or-after quirk to prevent the behavior
change.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::canChangeModuleScriptsExecutionTiming):
(WebCore::HTMLDocumentParser::notifyFinished):

Canonical link: https://commits.webkit.org/301665@main
